web/source/osk/bannerManager.ts

```diff
@@ -0,0 +1,177 @@
+      persistentBanner: false,
+      enablePredictions: true,
+      imagePath: ""
+    }
+
+    constructor() {
+      // Step 1 - establish the container element.  Must come before this.setOptions.
+      this.constructContainer();
+
+      // Initialize with the default options - any 'manually set' options come post-construction.
+      // This will also automatically set the default banner in place.
+      this.setOptions(BannerManager.DEFAULT_OPTIONS);
+
+      // Register a listener for model change events so that we can hot-swap the banner as needed.
+      let keyman = com.keyman.singleton;
+      keyman.modelManager['addEventListener']('modelchange', this.selectBanner.bind(this));
+    }
+
+    private constructContainer(): HTMLDivElement {
+      let height = 40;
+
+      let keymanweb = com.keyman.singleton;
+      let util = keymanweb.util;
+
+      let d = util._CreateElement('div');
+      d.id = "keymanweb_banner_container";
+      d.className = "keymanweb-banner-container";
+      return this.bannerContainer = d;
+    }
+
+    public get element(): HTMLDivElement {
+      return this.bannerContainer;
+    }
+
+    public getOptions(): BannerOptions {
+      let retObj = {};
+
+      for(let key in this._options) {
+        retObj[key] = this._options[key];
+      }
+
+      return retObj;
+    }
+
+    public setOptions(optionSpec: BannerOptions) {
+      let keyman = com.keyman.singleton;
+
+      for(let key in optionSpec) {
+        switch(key) {
+          // Each defined option may require specialized handling.
+          case 'persistentBanner':
+            // Determines the banner type to activate.
+            this.alwaysShows = optionSpec['persistentBanner'];
+            break;
+          case 'enablePredictions':
+            // If we're not going to show suggestions, it's best to turn off predictions.
+            keyman.modelManager.enabled = optionSpec['enablePredictions'];
+            break;
+          case 'imagePath':
+            // Determines the image file to use for ImageBanners.
+            this.imagePath = optionSpec['imagePath'];
+            break;
+          default:
+            // Invalid option specified!
+        }
+
+        this._options[key] = optionSpec['key'];
+      }
+
+      this.selectBanner();
+    }
+
+    public appendStyles() {
+      if(this.activeBanner) {
+        this.activeBanner.appendStyleSheet();
+      }
+    }
+
+    public setBanner(type: BannerType) {
+      switch(type) {
+        case 'blank':
+          this._setBanner(new BlankBanner());
+          break;
+        case 'image':
+          this._setBanner(new ImageBanner(this.imagePath));
+          break;
+        case 'suggestion':
+          this._setBanner(new SuggestionBanner());
+          let keyman = com.keyman.singleton;
+          let banner = this.activeBanner as SuggestionBanner;
+          keyman.modelManager['addEventListener']('invalidatesuggestions', banner.invalidateSuggestions.bind(this));
```

Comment:
How are these event listeners removed if the banner is replaced?

Reply:
@jahorton plans to add removing event listeners in a future PR